lib/FederationService.php

```diff
@@ -0,0 +1,215 @@
+		$serverHost = $this->urlGenerator->getAbsoluteURL('/');
+		$remoteFileUrl = \rtrim($server, '/') . '/index.php/apps/richdocuments/documents.php/federated' .
+			'?shareToken=' . $shareToken .
+			'&shareRelativePath=' . $shareRelativePath .
+			'&server=' . \rtrim($serverHost, '/') .
+			'&accessToken=' . $accessToken;
+		return $remoteFileUrl;
+	}
+	
+	/**
+	*
+	* @param string $server addres of a remote server
+	* @param string $accessToken wopi access token from a remote server
+	* @return array|null with additional wopi information
+	*/
+	public function getWopiForToken($server, $accessToken) {
+		$remote = $server;
+
+		if (!$this->isServerAllowed($remote)) {
+			$this->logger->info("Server {server} is not allowed.", ["server" => $remote]);
+			return null;
+		}
+
+		try {
+			$client = $this->httpClient->newClient();
+			$url = $remote . '/ocs/v2.php/apps/richdocuments/api/v1/federation';
+
+			$response = $client->post(
+				$url,
+				[
+					'form_params' => [
+						'token' => $accessToken,
+						'format' => 'json'
+					],
+					'timeout' => 3,
+					'connect_timeout' => 3,
+				]
+			);
+
+			$responseBody = $response->getBody();
+			$data = \json_decode($responseBody, true, 512);
+			if (\is_array($data)) {
+				return $data['ocs']['data'];
+			}
+			return null;
+		} catch (\Throwable $e) {
+			$this->logger->info('Cannot get the wopi info from remote server: ' . $remote, ['exception' => $e]);
+		}
+
+		return null;
+	}
+
+	/**
+	 * Check if server is allowed
+	 *
+	 * @param string $remote a remote url
+	 * @return bool indicating if given remote is allowed server
+	 */
+	public function isServerAllowed($remote) {
+		// TODO: implement check for trusted server, for a moment all trusted
+
+		return true;
+	}
+
+	/**
+	 * Get the wopiSrc Url from a remote server.
+	 *
+	 * @param string $server a remote
+	 * @return string with the wopi src Url
+	 */
+	public function getRemoteWopiSrc($server) {
+		if (\strpos($server, 'http://') === false && \strpos($server, 'https://') === false) {
+			$server = 'https://' . $server;
+		}
+
+		if (!$this->isServerAllowed($server)) {
+			$this->logger->info("Server {server} is not allowed.", ["server" => $server]);
+			return '';
+		}
+
+		try {
+			$getWopiSrcUrl = $server . '/ocs/v2.php/apps/richdocuments/api/v1/federation?format=json';
+			$client = $this->httpClient->newClient();
+			$response = $client->get($getWopiSrcUrl, ['timeout' => 5]);
+			$data = \json_decode($response->getBody(), true);
+
+			if (\is_array($data)) {
+				return $data['ocs']['data']['wopi_url'];
+			}
+		} catch (\Throwable $e) {
+			$this->logger->info('Cannot get the wopiSrc of remote server: ' . $server, ['exception' => $e]);
+		}
+
+		return '';
+	}
+
+	/**
+	 * split user and remote from federated cloud id, null if not federated cloud id
+	 *
+	 * @param string $userId user id
+	 * @return string
+	 */
+	public function generateFederatedCloudID(string $userId) : string {
+		if (\strpos($userId, '@') === false) {
```

Comment:
@jvillafanez this is copy of code from other federated codepaths. I would reference them here. These all, in case of changes, should be kept in sync. This logic is handled below more in detail.

Reply:
I can maybe see whether I can copy some tests..

Reply:
If it's copied from another place maybe we can include a reference, specially if the checks should be in sync, otherwise I could touch this piece only here without knowing.

Reply:
You were right here, because it contained logic that was in fact not really needed. when use clicks on federated share mount, owner is always federated user, and editor is always local user. Thus when generating token I just need to make editoruid being a federated user (as session is redirected to federated server)